### PR TITLE
[mlir][python] python binding wrapper for the affine.AffineForOp

### PR DESCRIPTION
**Pull request by kaitingwang**
This PR creates the wrapper class AffineForOp and adds a testcase for it.   A testcase for the AffineLoadOp is also added.

### CODE REVIEW

**Comment by makslevental** [comment]
Hey thanks @kaitingwang. Can you take a look here https://github.com/llvm/llvm-project/blob/baf42cdea2f243a1c9d2f89c456589e351c50521/mlir/python/mlir/dialects/scf.py#L106 and see if you can add the same thing for affine?

**Comment by kaitingwang** [comment]
> Hey thanks @kaitingwang. I'll take a look soon but can you take a look here
> 
> https://github.com/llvm/llvm-project/blob/baf42cdea2f243a1c9d2f89c456589e351c50521/mlir/python/mlir/dialects/scf.py#L106
> and see if you can add the same thing for affine?

Thanks for the suggestion!  I added the syntatic suger form of affine.for_ as well as some testcases (mostly adapted from scf.for)

File: mlir/python/mlir/dialects/affine.py
Comment (makslevental) [inline mlir/python/mlir/dialects/affine.py:71]:
```suggestion
            lowerBoundOperands=_get_op_results_or_values(lower_bound_operands),
```
I think this should work - same for `upperBoundOperands`.

File: mlir/test/python/dialects/affine.py
Comment (makslevental) [inline mlir/test/python/dialects/affine.py:16]: 👍

**Review by makslevental — APPROVED** [review]
LGTM modulo that one suggestion (not necessary ofc). Well organized code too. Thanks!

File: mlir/python/mlir/dialects/affine.py
Comment (kaitingwang) [reply mlir/python/mlir/dialects/affine.py]: Thanks!  Yes, it does work.  I changed it and pushed the change.

**Comment by kaitingwang** [comment]
CI was all green so I merged the PR.  Thanks @makslevental for the review!

File: mlir/python/mlir/dialects/affine.py
Comment (ftynse) [inline mlir/python/mlir/dialects/affine.py:6]: Nit: why do we need to import `AffineForOp` if we import `*` above?  (This is different for `_Dialect` that is private due to underscore in the name and is therefore not imported as part of `*`.)

File: mlir/python/mlir/dialects/affine.py
Comment (ftynse) [inline mlir/python/mlir/dialects/affine.py:24]: Nit: terminate Python comments with a full stop. Same below for some longer descriptions.

File: mlir/python/mlir/dialects/affine.py
Comment (ftynse) [inline mlir/python/mlir/dialects/affine.py:34]: These would _really_ benefit from having type annotations.

File: mlir/python/mlir/dialects/affine.py
Comment (ftynse) [inline mlir/python/mlir/dialects/affine.py:73]: It would be nice to support integers as lower/upper bound and have maps constructed for them on-the-fly.

File: mlir/python/mlir/dialects/affine.py
Comment (ftynse) [inline mlir/python/mlir/dialects/affine.py:112]: We should be doing this in the "main" builder, not in the wrapper.

File: mlir/python/mlir/dialects/affine.py
Comment (ftynse) [inline mlir/python/mlir/dialects/affine.py:124]: There is no need to create a constant operation and feed as a symbol into an affine map. Affine maps can have constant expressions and we should use that. I also suspect canonicalization or constant folding will immediately remove those constants.

File: mlir/test/python/dialects/affine.py
Comment (ftynse) [inline mlir/test/python/dialects/affine.py:120]: This looks rather confusing/scary to me...

**Review by ftynse — COMMENTED** [review]
I would have requested changes on this one. Please address promptly in a follow-up PR.

File: mlir/test/python/dialects/affine.py
Comment (makslevental) [reply mlir/test/python/dialects/affine.py:120]: Do you mean the `range = affine.for_`?  In the `scf` test I have `range_` (easy to have missed the trailing underscore...).

File: mlir/python/mlir/dialects/affine.py
Comment (makslevental) [reply mlir/python/mlir/dialects/affine.py:6]: I think explicitly importing names used in the file is not a bad thing. I also think this is the natural result of essentially treating these files as `__init__.py`s for modules that don't exist structurally (i.e., it should be `affine/__init__.py`) _and_ as implementation files as well. Anyway I'll take it out.

**Comment by makslevental** [comment]
WIP https://github.com/llvm/llvm-project/pull/74495

**Comment by kaitingwang** [comment]
@makslevental 

> WIP #74495.

Really appreciate your help to address @ftynse 's comments (next time I'll wait for your approval too)!  Thank you for your reviews.   I'll go over the left-over situation after work today and will be happy to address any remaining issues needed.  (At @makslevental's fast speed, there may not be anything left for me to address!)  :)

**Pull request merged**
